### PR TITLE
bpo-35899: Support of empty and weird strings by Enum

### PR DESCRIPTION
**Pull request by matrixise**
Co-authored-by: Maxwell <maxwellpxt@gmail.com>
Co-authored-by: Stéphane Wirtel <stephane@wirtel.be>

https://bugs.python.org/issue35899

### CODE REVIEW

File: Lib/test/test_enum.py [hunk 1]
```diff
@@ -2730,6 +2730,13 @@ def cycle_enum():
         self.assertEqual(256, len(seen), 'too many composite members created')
 
 
+class TestWeirdString(unittest.TestCase):
+    def test_weird_character(self):
+        Yay = Enum('Yay', ('!', 'B', 'C'))
```
Comment (tirkarthi) [inline Lib/test/test_enum.py:2740]: There should be a test for the original report where empty string causes an error like `Yay = Enum('Yay', ('', 'B', 'C'))` in https://bugs.python.org/issue35899#msg334866

**Review by bdbaraban — CHANGES_REQUESTED** [review]
Building on @tirkarthi's comment, I'd like to offer alternative tests that would test both empty and "weird" strings while also aligning more with the existing unittest layout in `test_enum.py`. I propose the following unittest functions to be entered starting at line 762 in the original `test_enum.py` file:

```
def test_programmatic_function_string_list_with_empty_string(self):
     SummerMonth = Enum('SummerMonth', ['june', 'july', ''])
     lst = list(SummerMonth)
     self.assertEqual(len(lst), len(SummerMonth))
     self.assertEqual(len(SummerMonth), 3, SummerMonth)
     self.assertEqual(
             [SummerMonth.june, SummerMonth.july, getattr(SummerMonth, '')],
             lst
             )
    for i, month in enumerate(['june', 'july', ''], 1):
        e = SummerMonth(i)
        self.assertEqual(int(e.value), i)
        self.assertNotEqual(e, i)
        self.assertEqual(e.name, month)
        self.assertIn(e, SummerMonth)
        self.assertIs(type(e), SummerMonth)
 
def test_programmatic_function_string_list_with_weird_string(self):
    SummerMonth = Enum('SummerMonth', ['june', 'july', '!'])
    lst = list(SummerMonth)
    self.assertEqual(len(lst), len(SummerMonth))
    self.assertEqual(len(SummerMonth), 3, SummerMonth)
    self.assertEqual(
            [SummerMonth.june, SummerMonth.july, getattr(SummerMonth, '!')],
            lst
            )
    for i, month in enumerate('june july !'.split(), 1):
        e = SummerMonth(i)
        self.assertEqual(int(e.value), i)
        self.assertNotEqual(e, i)
        self.assertEqual(e.name, month)
        self.assertIn(e, SummerMonth)
        self.assertIs(type(e), SummerMonth)
```

**Comment by matrixise** [comment]
Hi @bdbaraban 

1. You have to use the subTest() method when you want to create a test in a loop.
2. Your test is really complete but the other tests (from test_enum.py) already cover all the features of Enum.
3. A test should be the simplest as possible.

Thank you

File: Lib/test/test_enum.py (hunk 1 above)
Comment (matrixise) [reply Lib/test/test_enum.py:2740]: Hi @tirkarthi 
Thank you for your feedback, I have just added this test.

**Comment by bdbaraban** [comment]
> 
> 
> Hi @bdbaraban
> 
>     1. You have to use the subTest() method when you want to create a test in a loop.
> 
>     2. Your test is really complete but the other tests (from test_enum.py) already cover all the features of Enum.
> 
>     3. A test should be the simplest as possible.
> 
> 
> Thank you

Understood, thank you for the feedback! For clarification purposes, I based these tests off the other `programmatic_function` tests defined in `TestEnum`. They also run multiple tests in a for loop, without using the `subTest()` method. Is this incorrect?

**Comment by matrixise** [comment]
@bdbaraban 
`unittest.TestCase.subTest` exists since 3.4, same thing for Enum.
the code of `programmatic_function` is not incorrect, but in the case of a failed test, you will start to debug the test and will try to find the variables/values with the bad domain. with a context (like subTest), you will have these variables when you will execute the tests.

For my part, I prefer to use subTest (or in the case of pytest, `@pytest.mark.parametrize`) because you will execute your test in a context and for the debugging, it's easiest. See https://docs.python.org/3/library/unittest.html?highlight=subtest#distinguishing-test-iterations-using-subtests

Now, when we write new code for Python, test or code, we don't use the new features of Python. For example, the [PEP 572](https://www.python.org/dev/peps/pep-0572/)  has been merged into CPython, we won't start to rewrite the library and the tests with the new syntax. It's the same with the f-string.

**Comment by matrixise** [comment]
@ethanfurman I close this PR because you can use the PR of @bdbaraban.

Have a nice day,

Stéphane

**Pull request closed without merging**